app/models/transformation_mapping/vm_migration_validator.rb

```diff
@@ -0,0 +1,144 @@
+class TransformationMapping::VmMigrationValidator
+  VM_CONFLICT = "conflict".freeze
+  VM_EMPTY_NAME = "empty_name".freeze
+  VM_IN_OTHER_PLAN = "in_other_plan".freeze
+  VM_INVALID = "invalid".freeze
+  VM_MIGRATED = "migrated".freeze
+  VM_NOT_EXIST = "not_exist".freeze
+  VM_VALID = "ok".freeze
+
+  def initialize(mapping, vm_list = nil)
+    @mapping = mapping
+    @vm_list = vm_list
+  end
+
+  def validate
+    @vm_list.present? ? identify_vms : select_vms
+  end
+
+  def select_vms
+    valid_list = []
+
+    @mapping.transformation_mapping_items.where(:source_type => EmsCluster).collect(&:source).collect(&:vms).flatten.each do |vm|
```

Comment:
Any performance concerns around the number of objects being loaded and the amount of ruby processing being done here?

Reply:
I am working on the performance improvement. Any suggestion is welcomed.

Reply:
It could be 2 queries with:
```ruby
cluster_ids = @mapping.transformation_mapping_items.where(:source_type => EmsCluster).pluck(:source_id)
Vm.where(:ems_cluster_id => cluster_ids).all.each do |vm|
.....
```
Maybe @kbrock or @NickLaMuro have other suggestions?